File: src/main/java/edu/cnm/deepdive/sieve/bitset/Sieve.java
package edu.cnm.deepdive.sieve.bitset;

import java.util.BitSet;

public class Sieve {

  private static BitSet sieve(int limit) {
    BitSet candidates = new BitSet(limit + 1);
    candidates.set(2, limit + 1);
    for (int prime = 2; prime <= Math.sqrt(limit); prime = candidates.nextSetBit(prime + 1)) {
      for (int multiple = prime * prime; multiple <= limit; multiple += prime) {
        candidates.clear(multiple);
      }
    }
    return candidates;
  }

  public static void main(String... args) {
    int upperBound = 1_000_000;
    long start = System.currentTimeMillis();
    BitSet primes = sieve(upperBound);
    long end = System.currentTimeMillis();
    System.out.printf("Java Sieve with BitSet:%n"
            + "%1$,d primes found between %2$,d and %3$,d in %4$d ms.%n",
        primes.cardinality(), primes.nextSetBit(0), primes.previousSetBit(primes.size() - 1),
        end - start);
  }

}

File: src/main/kotlin/edu/cnm/deepdive/sieve/bitset/sieve.kt
package edu.cnm.deepdive.sieve.bitset

import java.util.BitSet

fun sieve(limit: Int): BitSet {
    val primes = BitSet()
    primes.set(2, limit + 1)
    var prime = 2
    while (prime <= Math.sqrt(limit.toDouble()).toInt()) {
        for (multiple in (prime * prime)..limit step prime) {
            primes.clear(multiple)
        }
        prime = primes.nextSetBit(prime + 1)
    }
    return primes
}

fun main(args: Array<String>) {
    val start = System.currentTimeMillis()
    val upperBound = 1_000_000
    val primes = sieve(upperBound)
    val end = System.currentTimeMillis()
    println(
        """
            Kotlin Sieve with BitSet: 
            ${primes.cardinality()} primes found between ${primes.nextSetBit(0)} and ${primes.previousSetBit(primes.size() - 1)} (inclusive) in ${end - start} ms.
        """.trimIndent()
    )
}